Add instructions for wrapping new module (#1687)
GenericMappingTools/pygmt: doc/contributing.md

Reviewer: ```suggestion
When writing tests, don't test everything that the GMT function already tests, such as
every unique combination of arguments. An exception to this would be the most popular
```

Reviewer: ```suggestion
The highest priority for tests should be the Python-specific code, such as numpy,
pandas, and Xarray objects and the virtual file mechanism.
```

Reviewer: As we are not strictly following this rule, do we want to keep it in the guidelines?

    [Testing plots](#testing-plots).
  - Include an example of new features in the gallery or tutorials.
    Please refer to [Gallery plots](#gallery-plots) or [Tutorials](#tutorials).

>     If adding a new method/function/class, the gallery example or tutorial can

Reviewer 2: ```suggestion
    If adding a new method/function/class, the gallery example or tutorial should
```
I think it's better to suggest having the module examples in a different PR than the one that wraps them (which I think is the intent of previous guidelines).

>     be submitted in a separate pulll request.

Reviewer 2: ```suggestion
    be submitted in a separate pull request.
```
Spelling fix

make lint    # Runs pylint, which is a bit slower
```

### Wrapping a new GMT module

Wrapping a new GMT module in PyGMT is usually a big task, which will progress

> quicker and smoother if done in **small, maneable chunks**. This section

Reviewer 3: Is "maneable" a typo?

Author: Yes, thanks for catching it.

> quicker and smoother if done in **small, maneable chunks**. This section

Reviewer 2: ```suggestion
quicker and smoother if done in **small, manageable chunks**. This section
```
Per @seisman's comment

gives an overview of the specific tasks involved in wrapping a new module.

1. Create a ['Wrapper for `<module-name>`' feature request issue](#request-wrapper-for-gmt-module). 
2. Open a ['Wrap `<module-name>`' initial feature implementation PR](#initial-feature-implementation).
3. Open an ['Add missing aliases to `<module-name>`' documentation PR](#add-missing-aliases).
4. Open a 'support additional functionality in module' PR (optional).
5. Add ['Gallery example for module' documentation PR](#contributing-gallery-plots).
6. Add ['Tutorial for module' documentation PR](#contributing-tutorials) (optional).

These steps will be tracked in the 'Wrapper for `<module-name>`' issue and the
['wrapping GMT modules'](https://github.com/GenericMappingTools/pygmt/projects/9)

> project board. The pull requests can be split between multiple contributors and

Reviewer 2: I think we should either track the module updates in a project board (my preference) or an issue, but not both.

Author: Do you think we should abandon the 'Wrapper for `<module-name>`' issues entirely or keep the issues/issue-template and just move the checklist to the project board? I think the main benefit of the issues is that from my understanding any GitHub user can open/comment on an issue whereas writing on the project board requires special permissions.

Reviewer 2: I think the project board works better for tracking the progress of wrapping a module, and that having a separate issue for wrapping the function is a bit redundant. I see what you mean about the benefits of issues being open to everyone, but my thought is that wrapping modules is a pretty routine process and doesn't need to first be raised as an issue (unlike a bug or a feature request). We can add some/all of the GMT modules to the project board to track their process. Users who don't have write permission are still free to open up an issue requesting a module/feature (or a pull request! :crossed_fingers:) as a way to let us know that a certain feature would be useful to them.

Reviewer 3: > the benefits of issues being open to everyone

IMHO, issues are also more visible than project boards, so that we may have more potential contributors helping wrap new modules and add more aliases.

wrap a new GMT module (e.g., [initial grdfill implementation](https://github.com/GenericMappingTools/pygmt/pull/1276/files)):

* Create a new module `<module-name>.py` in `pygmt/src`. The module docstring
  should include the module name and a short description of the functionality
  (e.g., `grdfill - Fill blank areas from a grid.`).
* Add a function `<module-name>` to the module. When writing the new function,
  it is generally easiest to reference the source code for other functions that
  input/output similar object types.
* Add a detailed docstring following the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
* Add the function to the import statements in `pygmt/src/__init__.py`.
* Add the function to the import statements in `pygmt/__init__.py`.
* Add the function to appropriate section of the API documentation in `doc/api/index.rst`.
* Add a testing module `test_<module-name>.py` in `pygmt/tests`, following
  the guidelines in the [testing your code](#testing-your-code) section.

#### Add missing aliases

After the initial implementation, the missing aliases can be added in a

> separate PR (e.g., [add missing aliases to grd2xyz](https://github.com/GenericMappingTools/pygmt/pull/1537/files)).

Reviewer 4: Might need to get a newer example once we go with the long option standard in #1932.

* Select a suitable alias for each GMT option, following the guidelines in the
  [code style](#code-style) section. Before creating a new alias, check if the

>   parameter is listed in the `COMMON_OPTIONS` dictionary in

Reviewer 4: TODO once #1932 is merged.
```suggestion
  parameter is listed in the `COMMON_DOCSTRINGS` dictionary in
```

  `pygmt/helpers/decorators.py`, if other wrapped GMT modules have a similar
  parameter, and if [GMT.jl](https://www.generic-mapping-tools.org/GMT.jl/dev/)
  has defined an alias.
* Update the `use_alias` decorator for the `<module-name>` function using the
  GMT option as the parameter and the alias as the argument.
* Add the alias and description to the parameters section of the docstring,
  using the `fmt_docstring` decorator to add descriptions for parameters

>   included in the `COMMON_OPTIONS` dictionary.

Reviewer 4: ```suggestion
  included in the `COMMON_DOCSTRINGS` dictionary.
```

make check   # Runs ruff in check mode
```

### Wrapping a GMT Module

Wrapping a GMT module in PyGMT is usually a big task, but it can progress more smoothly
and efficiently when divided into **small, manageable chunks**. This section gives an
overview of the main tasks involved.

1. Create a feature request ("wrapper request issue") for wrapping a module and discuss

>    what users would like to see in the wrapper [optional, usually done by users].

Reviewer 4: ```suggestion
1. Open "wrapper request" issue - create a feature request for wrapping a module and
   discuss what features should be available in the wrapper [optional, usually done by
   users].
```

1. Open "wrapper request" issue - create a feature request for wrapping a module and
   discuss what features should be available in the wrapper [optional, usually done by
   users].
2. Open a "wrapper tracking" issue - use the "Wrapper for a GMT module" issue template,
   to track the progress of wrapping the module. Link it to the
   [Project board](https://github.com/orgs/GenericMappingTools/projects/3), and close
   the "wrapper request issue" with a comment such as [usually done by maintainers]:
   > Thank you for opening the feature request. The progress of wrapping the module will
   > be tracked in issue #XXX and
   > the [Project board](https://github.com/orgs/GenericMappingTools/projects/3).
3. Open one PR for the initial implementation, focusing on required and essential 
   parameters [done by maintainers or contributors].
4. Close the "wrapper tracking issue" once the initial implementation is merged. Leave a
   comment such as [done by maintainers]:
   > The initial implementation of wrapping the XXX module was completed in PR #XXX.
   > Not all functionalities are implemented yet. Further progress will be tracked in
   > the Project board.

>    This is necessary to avoid having hundreds of long-term open issues.

Reviewer 4: I only counted 159 from https://docs.generic-mapping-tools.org/6.6/modules.html so not hundreds, but still a lot :laughing:

```suggestion
   This is necessary to avoid having too many long-term open issues.
```

6. Open one PR to add a gallery example or a tutorial.

These PRs can be split among multiple contributors. There is no obligation for a single
contributor to complete all steps. Please comment on the "wrapper tracking issue" if you
would like to open a PR for any of these tasks to avoid redundant efforts.

#### Initial Feature Implementation

First, comment on the "Wrapper tracking issue" that you will be working on the initial
implementation. This first pull request should be as minimal as possible - only adding
the required functionality (i.e., wrapping the required GMT parameters and supporting
the primary input/output types).

The following steps are common to all initial implementation pull requests that wrap a
GMT module:

* Create a new module `<module-name>.py` in `pygmt/src`. The module docstring should
  include the module name and a short description of the functionality (e.g.,
  `grdfill - Fill blank areas from a grid.`).
* Add a function `<module-name>` to the module. When writing the new function, it is
  generally easiest to reference the source code for other functions that input/output
  similar object types.

> * Write a detailed docstring following the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).

Reviewer 4: Wrap around to avoid extra long line.

```suggestion
* Write a detailed docstring following the
  [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
```

* Add the function to the import statements in `pygmt/src/__init__.py`.

> * Add the function to the import statements in `pygmt/__init__.py`.

Reviewer 4: Combine
```suggestion
* Add the function to the import statements in `pygmt/src/__init__.py` and
  `pygmt/__init__.py`.
```

   > Not all functionalities are implemented yet. Further progress will be tracked in
   > the Project board.
   This is necessary to avoid having hundreds of long-term open issues.
5. Open one or more PRs to implement the remaining features and missing aliases.
6. Open one PR to add a gallery example or a tutorial.

These pull requests can be split among multiple contributors. There is no obligation for
a single contributor to complete all steps. Please comment on the "wrapper tracking
issue" if you would like to open a pull request for any of these tasks to avoid
redundant efforts.

#### Feature Request Issue for Wrapping a GMT Module

* Find the [*Issues*](https://github.com/GenericMappingTools/pygmt/issues) tab at the
  top of the GitHub repository and click *New Issue*.
* Choose the *Feature request* issue template.
* Follow the prompts for filling out the issue template.

#### Central Issue for Tracking the Progress

*This step is usually done by maintainers.*

* Find the [*Issues*](https://github.com/GenericMappingTools/pygmt/issues) tab on the
  top of the GitHub repository and click *New Issue*.
* Choose the "Wrapper for a module" issue template.

> * Fill out the issue template.

Reviewer 3: Maybe remove these two sections, since they're already covered in https://pygmt-dev--1687.org.readthedocs.build/en/1687/contributing.html#submitting-a-feature-request.

the required functionality (i.e., wrapping the required GMT parameters and supporting
the primary input/output types).

The following steps are common to all initial implementation pull requests that wrap a
GMT module:

* Create a new module `<module-name>.py` in `pygmt/src`. The module docstring should
  include the module name and a short description of the functionality (e.g.,
  `grdfill - Fill blank areas from a grid.`).
* Add a function `<module-name>` to the module. When writing the new function, it is
  generally easiest to reference the source code for other functions that input/output
  similar object types.
* Write a detailed docstring following the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
* Add the function to the import statements in `pygmt/src/__init__.py`.
* Add the function to the import statements in `pygmt/__init__.py`.
* Add the function to appropriate section of the API documentation in `doc/api/index.rst`.
* Add a testing module `test_<module-name>.py` in `pygmt/tests`, following
  the guidelines in the [testing your code](#testing-your-code) section.

> #### Add missing aliases

Reviewer: ```suggestion
#### Add Missing Aliases
```
